### ISSUE TITLE
[bitnami/postgresql-repmgr] Witness container can't work with primary/secondary node through any other port but 5432

### ISSUE THREAD

**Issue by hillout** [bitnami/containers#69471]
### Name and Version

bitnami/postgresql-repmgr:latest

### What architecture are you using?

amd64

### What steps will reproduce the bug?

Let's say, for postgres I need to use any other TCP port apart from `5432`. So, here is my example:

```
mkdir -p data/pg0 && mkdir -p data/pg1 && mkdir -p data/pgw0 && chown 1001 -R data
touch .env && nano .env
```

Put `PG_PORT=5445` in the `.env` file.

docker-compose.yml file:

```yaml
services:
  pg-0:
    image: bitnami/postgresql-repmgr:latest
    env_file:
      - .env    
    volumes:
      - ./data/pg0:/bitnami/postgresql
    environment:
      - POSTGRESQL_POSTGRES_PASSWORD=adminpassword
      - POSTGRESQL_USERNAME=customuser
      - POSTGRESQL_PASSWORD=custompassword
      - POSTGRESQL_DATABASE=customdatabase
      - POSTGRESQL_PORT_NUMBER=${PG_PORT}      
      - REPMGR_PASSWORD=repmgrpassword
      - REPMGR_PRIMARY_HOST=pg-0
      - REPMGR_PRIMARY_PORT=${PG_PORT}
      - REPMGR_PARTNER_NODES=pg-0:${PG_PORT},pg-1:${PG_PORT},pgw-0:${PG_PORT}
      - REPMGR_NODE_NAME=pg-0
      - REPMGR_NODE_NETWORK_NAME=pg-0
      - REPMGR_PORT_NUMBER=${PG_PORT}
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres -p ${PG_PORT}"]
      interval: 10s
      timeout: 5s
      retries: 5

  pg-1:
    image: bitnami/postgresql-repmgr:latest
    env_file:
      - .env    
    volumes:
      - ./data/pg1:/bitnami/postgresql
    environment:
      - POSTGRESQL_POSTGRES_PASSWORD=adminpassword
      - POSTGRESQL_USERNAME=customuser
      - POSTGRESQL_PASSWORD=custompassword
      - POSTGRESQL_DATABASE=customdatabase
      - POSTGRESQL_PORT_NUMBER=${PG_PORT}      
      - REPMGR_PASSWORD=repmgrpassword
      - REPMGR_PRIMARY_HOST=pg-0
      - REPMGR_PRIMARY_PORT=${PG_PORT}
      - REPMGR_PARTNER_NODES=pg-0:${PG_PORT},pg-1:${PG_PORT},pgw-0:${PG_PORT}
      - REPMGR_NODE_NAME=pg-1
      - REPMGR_NODE_NETWORK_NAME=pg-1
      - REPMGR_PORT_NUMBER=${PG_PORT}
    depends_on:
      pg-0:
        condition: service_healthy
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres -p ${PG_PORT}"]
      interval: 10s
      timeout: 5s
      retries: 5
          
  pgw-0:
    image: bitnami/postgresql-repmgr:latest
    env_file:
      - .env    
    volumes:
      - ./data/pgw0:/bitnami/postgresql
    environment:
      - POSTGRESQL_POSTGRES_PASSWORD=adminpassword
      - POSTGRESQL_USERNAME=customuser
      - POSTGRESQL_PASSWORD=custompassword
      - POSTGRESQL_DATABASE=customdatabase
      - POSTGRESQL_PORT_NUMBER=${PG_PORT}
      - REPMGR_PASSWORD=repmgrpassword
      - REPMGR_PRIMARY_HOST=pg-0
      - REPMGR_PRIMARY_PORT=${PG_PORT}
      - REPMGR_PARTNER_NODES=pg-0:${PG_PORT},pg-1:${PG_PORT},pgw-0:${PG_PORT}
      - REPMGR_NODE_NAME=pgw-0
      - REPMGR_NODE_NETWORK_NAME=pgw-0
      - REPMGR_PORT_NUMBER=${PG_PORT}
      - REPMGR_NODE_TYPE=witness
      - REPMGR_NODE_ID_START_SEED=2000
      - BITNAMI_DEBUG=true
    depends_on:
      pg-0:
        condition: service_healthy
      pg-1:
        condition: service_healthy
```

pg0 and pg1 are working together great, in the logs of pg0 I see:
`[2024-07-23 10:57:49] [NOTICE] new standby "pg-1" (ID: 1001) has connected`

But the witness is not connecting and from extended logs we can see that after some attempts it's trying to use port 5432 to connect to pg-0 for some reason:

```
postgresql-repmgr 10:57:55.38 INFO  ==> Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 10:57:55.38 INFO  ==> Subscribe to project updates by watching https://github.com/bitnami/containers
postgresql-repmgr 10:57:55.38 INFO  ==> Submit issues and feature requests at https://github.com/bitnami/containers/issues
postgresql-repmgr 10:57:55.39 INFO  ==> Upgrade to Tanzu Application Catalog for production environments to access custom-configured and pre-packaged software components. Gain enhanced features, including Software Bill of Materials (SBOM), CVE scan result reports, and VEX documents. To learn more, visit https://bitnami.com/enterprise
postgresql-repmgr 10:57:55.39 INFO  ==> 
postgresql-repmgr 10:57:55.40 DEBUG ==> Configuring libnss_wrapper...
postgresql-repmgr 10:57:55.42 DEBUG ==> Copying files from /opt/bitnami/postgresql/conf.default to /opt/bitnami/postgresql/conf
postgresql-repmgr 10:57:55.42 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 10:57:55.45 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 10:57:55.45 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 10:57:55.45 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 10:57:55.46 DEBUG ==> Checking node 'pg-0:5445'...
postgresql-repmgr 10:57:55.49 DEBUG ==> Pretending primary role node - 'pg-0:5445'
postgresql-repmgr 10:57:55.49 DEBUG ==> Pretending primary set to 'pg-0:5445'!
postgresql-repmgr 10:57:55.50 DEBUG ==> Checking node 'pg-1:5445'...
postgresql-repmgr 10:57:55.53 DEBUG ==> Pretending primary role node - 'pg-0:5445'
postgresql-repmgr 10:57:55.54 DEBUG ==> Checking node 'pgw-0:5445'...
psql: error: connection to server at "pgw-0" (192.168.0.4), port 5445 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
postgresql-repmgr 10:57:55.55 DEBUG ==> Skipping: failed to get primary from the node 'pgw-0:5445'!
postgresql-repmgr 10:57:55.55 INFO  ==> Auto-detected primary node: 'pg-0:5445'
postgresql-repmgr 10:57:55.56 DEBUG ==> Primary node: 'pg-0:5445'
postgresql-repmgr 10:57:55.56 INFO  ==> Node configured as witness
postgresql-repmgr 10:57:55.57 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 10:57:55.57 DEBUG ==> Injecting a new postgresql.conf file...
postgresql-repmgr 10:57:55.57 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 10:57:55.63 DEBUG ==> Injecting a new pg_hba.conf file...
postgresql-repmgr 10:57:55.64 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 10:57:55.65 DEBUG ==> Node ID: '2000', Rol: 'witness', Primary Node: 'pg-0:5445'
postgresql-repmgr 10:57:55.65 INFO  ==> Initializing Repmgr...
postgresql-repmgr 10:57:55.65 INFO  ==> Initializing PostgreSQL database...
postgresql-repmgr 10:57:55.65 DEBUG ==> Copying files from /bitnami/postgresql/conf to /opt/bitnami/postgresql/conf
postgresql-repmgr 10:57:55.66 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql-repmgr 10:57:55.66 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql-repmgr 10:57:55.66 DEBUG ==> Ensuring expected directories/files exist...
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.UTF-8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
fixing permissions on existing directory /bitnami/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
syncing data to disk ... ok
Success. You can now start the database server using:
    /opt/bitnami/postgresql/bin/pg_ctl -D /bitnami/postgresql/data -l logfile start
postgresql-repmgr 10:57:56.39 INFO  ==> Starting PostgreSQL in background...
waiting for server to start....2024-07-23 10:57:56.417 GMT [173] LOG:  pgaudit extension initialized
2024-07-23 10:57:56.425 GMT [173] LOG:  redirecting log output to logging collector process
2024-07-23 10:57:56.425 GMT [173] HINT:  Future log output will appear in directory "/opt/bitnami/postgresql/logs".
2024-07-23 10:57:56.425 GMT [173] LOG:  starting PostgreSQL 16.3 on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2024-07-23 10:57:56.426 GMT [173] LOG:  listening on IPv6 address "::1", port 5445
2024-07-23 10:57:56.426 GMT [173] LOG:  listening on IPv4 address "127.0.0.1", port 5445
2024-07-23 10:57:56.428 GMT [173] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5445"
2024-07-23 10:57:56.432 GMT [177] LOG:  database system was shut down at 2024-07-23 10:57:56 GMT
2024-07-23 10:57:56.437 GMT [173] LOG:  database system is ready to accept connections
 done
server started
CREATE DATABASE
postgresql-repmgr 10:57:56.57 INFO  ==> Changing password of postgres
ALTER ROLE
postgresql-repmgr 10:57:56.59 INFO  ==> Creating user customuser
CREATE ROLE
postgresql-repmgr 10:57:56.61 INFO  ==> Granting access to "customuser" to the database "customdatabase"
GRANT
ALTER DATABASE
postgresql-repmgr 10:57:56.63 INFO  ==> Setting ownership for the 'public' schema database "customdatabase" to "customuser"

ALTER SCHEMA
postgresql-repmgr 10:57:56.65 INFO  ==> Creating replication user repmgr
CREATE ROLE
postgresql-repmgr 10:57:56.67 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down....2024-07-23 10:57:56.677 GMT [173] LOG:  received fast shutdown request
2024-07-23 10:57:56.679 GMT [173] LOG:  aborting any active transactions
2024-07-23 10:57:56.681 GMT [173] LOG:  background worker "logical replication launcher" (PID 181) exited with exit code 1
2024-07-23 10:57:56.681 GMT [175] LOG:  shutting down
2024-07-23 10:57:56.702 GMT [175] LOG:  checkpoint starting: shutdown immediate
2024-07-23 10:57:56.814 GMT [175] LOG:  checkpoint complete: wrote 932 buffers (5.7%); 0 WAL file(s) added, 0 removed, 1 recycled; write=0.020 s, sync=0.013 s, total=0.113 s; sync files=308, longest=0.002 s, average=0.001 s; distance=11353 kB, estimate=11353 kB; lsn=0/2000028, redo lsn=0/2000028
2024-07-23 10:57:56.828 GMT [173] LOG:  database system is shut down
 done
server stopped
postgresql-repmgr 10:57:56.89 INFO  ==> Configuring replication parameters
postgresql-repmgr 10:57:56.92 INFO  ==> Configuring fsync
postgresql-repmgr 10:57:56.93 INFO  ==> Starting PostgreSQL in background...
waiting for server to start....2024-07-23 10:57:56.962 GMT [252] LOG:  pgaudit extension initialized
2024-07-23 10:57:56.970 GMT [252] LOG:  redirecting log output to logging collector process
2024-07-23 10:57:56.970 GMT [252] HINT:  Future log output will appear in directory "/opt/bitnami/postgresql/logs".
2024-07-23 10:57:56.970 GMT [252] LOG:  starting PostgreSQL 16.3 on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2024-07-23 10:57:56.971 GMT [252] LOG:  listening on IPv4 address "0.0.0.0", port 5445
2024-07-23 10:57:56.971 GMT [252] LOG:  listening on IPv6 address "::", port 5445
2024-07-23 10:57:56.973 GMT [252] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5445"
2024-07-23 10:57:56.977 GMT [256] LOG:  database system was shut down at 2024-07-23 10:57:56 GMT
2024-07-23 10:57:56.982 GMT [252] LOG:  database system is ready to accept connections
 done
server started
postgresql-repmgr 10:57:57.06 INFO  ==> Creating repmgr user: repmgr
ERROR:  role "repmgr" already exists
2024-07-23 10:57:57.083 GMT [267] ERROR:  role "repmgr" already exists
2024-07-23 10:57:57.083 GMT [267] STATEMENT:  CREATE ROLE "repmgr" WITH LOGIN CREATEDB PASSWORD 'repmgrpassword';
ALTER ROLE
ALTER ROLE
postgresql-repmgr 10:57:57.12 INFO  ==> Creating repmgr database: repmgr
CREATE DATABASE
postgresql-repmgr 10:57:57.19 INFO  ==> Unregistering witness node...
NOTICE: using provided configuration file "/opt/bitnami/repmgr/conf/repmgr.conf"
2024-07-23 10:57:57.212 GMT [283] ERROR:  relation "repmgr.nodes" does not exist at character 108
2024-07-23 10:57:57.212 GMT [283] STATEMENT:    SELECT node_id, conninfo,          CASE WHEN type = 'primary' THEN 1 ELSE 2 END AS type_priority	   FROM repmgr.nodes    WHERE active IS TRUE      AND type != 'witness' ORDER BY active DESC, type_priority, priority, node_id
ERROR: _get_primary_connection(): unable to retrieve node records
DETAIL: 
ERROR:  relation "repmgr.nodes" does not exist
LINE 1: ...imary' THEN 1 ELSE 2 END AS type_priority    FROM repmgr.nod...
                                                             ^
DETAIL: query text is:
  SELECT node_id, conninfo,          CASE WHEN type = 'primary' THEN 1 ELSE 2 END AS type_priority	   FROM repmgr.nodes    WHERE active IS TRUE      AND type != 'witness' ORDER BY active DESC, type_priority, priority, node_id
ERROR: unable to connect to primary
DETAIL: 
connection pointer is NULL
postgresql-repmgr 10:57:57.21 INFO  ==> Registering witness node...
postgresql-repmgr 10:57:57.21 INFO  ==> Waiting for primary node...
postgresql-repmgr 10:57:57.21 DEBUG ==> Wait for schema repmgr.repmgr on 'pg-0:5445', will try 6 times with 10 delay seconds (TIMEOUT=60)
postgresql-repmgr 10:57:57.24 DEBUG ==> Schema repmgr.repmgr exists!
NOTICE: using provided configuration file "/opt/bitnami/repmgr/conf/repmgr.conf"
ERROR: connection to database failed
DETAIL: 
connection to server at "pg-0" (192.168.0.2), port 5432 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
ERROR: unable to connect to the primary node
HINT: a primary node must be configured before registering a witness node
postgresql-repmgr 10:57:57.26 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down....2024-07-23 10:57:57.269 GMT [252] LOG:  received fast shutdown request
2024-07-23 10:57:57.270 GMT [252] LOG:  aborting any active transactions
2024-07-23 10:57:57.272 GMT [252] LOG:  background worker "logical replication launcher" (PID 260) exited with exit code 1
2024-07-23 10:57:57.272 GMT [254] LOG:  shutting down
2024-07-23 10:57:57.303 GMT [254] LOG:  checkpoint starting: shutdown immediate
2024-07-23 10:57:57.337 GMT [254] LOG:  checkpoint complete: wrote 927 buffers (5.7%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.021 s, sync=0.011 s, total=0.036 s; sync files=304, longest=0.002 s, average=0.001 s; distance=16384 kB, estimate=16384 kB; lsn=0/3000028, redo lsn=0/3000028
2024-07-23 10:57:57.346 GMT [252] LOG:  database system is shut down
 done
server stopped
```

But when you change the port to 5432 in the `.env` file, it's working back again: 

```
postgresql-repmgr 11:03:52.29 INFO  ==> ** Starting repmgrd **
[2024-07-23 11:03:52] [NOTICE] repmgrd (repmgrd 5.4.1) starting up
INFO:  set_repmgrd_pid(): provided pidfile is /tmp/repmgrd.pid
[2024-07-23 11:03:52] [NOTICE] starting monitoring of node "pgw-0" (ID: 2000)
```

### What is the expected behavior?

I can use whatever port I want. 

### What do you see instead?

There is an error and witness is not connecting.

**Comment by monixca345** [comment]
That’s what I was looking for, what an info present here at this website, thank you admin!

**Comment by rafariossaa** [comment]
Hi,
 I was able to reproduce the issue. I am creating a fix for it. The problem lies in functions `repmgr_unregister_witness` and `repmgr_register_witness` that are not using the port environment variable.

**Comment by rafariossaa** [comment]
Hi,
  Could you give a try to the new release `bitnami/postgresql-repmgr:16.3.0-debian-12-r19` ?

**Comment by hillout** [comment]
@rafariossaa hi! 

It's wirking now, I used port 5445.
```
[2024-07-30 08:31:58] [NOTICE] new standby "pg-1" (ID: 1001) has connected
[2024-07-30 08:32:10] [NOTICE] new witness "pgw-0" (ID: 2000) has connected
[2024-07-30 08:32:16] [NOTICE] new witness "pgw-0" (ID: 2000) has connected
```
Seems now everything's OK. I turned off pg0, pg1 became primary and witness switched to it without a problem.

If I want to change the port on the fly is there any other option to do it? 

I've found the way:
`docker run --rm -v ./backup/latest:/backup dockerhub/mirror/postgres:16.0-bookworm \
  bash -c "PGPASSWORD={repmgr_password} psql -h 172.21.5.6 -p 5432 -U repmgr repmgr -c \"UPDATE repmgr.nodes SET conninfo = REPLACE(conninfo, 'port=5445', 'port=5432');\""`
